server/schedulers/random_split.go

```diff
@@ -0,0 +1,148 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schedulers
+
+import (
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pingcap/log"
+	"github.com/tikv/pd/pkg/errs"
+	"github.com/tikv/pd/server/core"
+	"github.com/tikv/pd/server/schedule"
+	"github.com/tikv/pd/server/schedule/filter"
+	"github.com/tikv/pd/server/schedule/operator"
+	"github.com/tikv/pd/server/schedule/opt"
+)
+
+const (
+	// RandomSplitName is random split scheduler name.
+	RandomSplitName = "random-split-scheduler"
+	// RandomSplitType is random split scheduler type.
+	RandomSplitType = "random-split"
+)
+
+func init() {
+	schedule.RegisterSliceDecoderBuilder(RandomSplitType, func(args []string) schedule.ConfigDecoder {
+		return func(v interface{}) error {
+			conf, ok := v.(*randomSplitSchedulerConfig)
+			if !ok {
+				return errs.ErrScheduleConfigNotExist.FastGenByArgs()
+			}
+			ranges, err := getKeyRanges(args)
+			if err != nil {
+				return err
+			}
+			conf.Ranges = ranges
+			conf.Name = RandomSplitName
+			return nil
+		}
+	})
+	schedule.RegisterScheduler(RandomSplitType, func(opController *schedule.OperatorController, storage *core.Storage, decoder schedule.ConfigDecoder) (schedule.Scheduler, error) {
+		conf := &randomSplitSchedulerConfig{}
+		if err := decoder(conf); err != nil {
+			return nil, err
+		}
+		return newRandomSplitScheduler(opController, conf), nil
+	})
+}
+
+type randomSplitSchedulerConfig struct {
+	Name   string          `json:"name"`
+	Ranges []core.KeyRange `json:"ranges"`
+}
+
+type randomSplitScheduler struct {
+	*BaseScheduler
+	conf    *randomSplitSchedulerConfig
+	filters []filter.Filter
+}
+
+// newRandomSplitScheduler creates an admin scheduler that randomly picks a region then splits it.
+// Splitting a region is not a heavy operation. Its cost is similar to transferring a leader, so we use the
+// same limit and config as shuffle-leader-scheduler.
+func newRandomSplitScheduler(opController *schedule.OperatorController, conf *randomSplitSchedulerConfig) schedule.Scheduler {
+	filters := []filter.Filter{
+		&filter.StoreStateFilter{ActionScope: conf.Name, TransferLeader: true},
+		filter.NewSpecialUseFilter(conf.Name),
+	}
+	base := NewBaseScheduler(opController)
+	return &randomSplitScheduler{
+		BaseScheduler: base,
+		conf:          conf,
+		filters:       filters,
+	}
+}
+
+func (s *randomSplitScheduler) GetName() string {
+	return s.conf.Name
+}
+
+func (s *randomSplitScheduler) GetType() string {
+	return RandomSplitType
+}
+
+func (s *randomSplitScheduler) EncodeConfig() ([]byte, error) {
+	return schedule.EncodeConfig(s.conf)
+}
+
+func (s *randomSplitScheduler) IsScheduleAllowed(cluster opt.Cluster) bool {
+	allowed := s.OpController.OperatorCount(operator.OpSplit) < cluster.GetOpts().GetLeaderScheduleLimit()
```

Comment:
Why is it limited by leader schedule limit?

Reply:
Can we configure OpSplit limit in scheduler config?